truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/EventListener.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2016, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.truffle.api.instrumentation;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+
+/**
+ * A listener attached by an {@link Instrumenter} using a {@link SourceSectionFilter} to specific
+ * locations of a guest language program. Listeners are shared across all source locations specified
+ * by the {@link SourceSectionFilter}. Use event listeners if your instrumentation does not need to
+ * store state per source location. If it is recommended to use a {@link EventNodeFactory} instead.
+ */
+public interface EventListener {
+
+    /**
+     * Invoked before an instrumented node is executed. The provided frame is the frame of
+     * instrumented node.
+     */
+    void onEnter(EventContext context, VirtualFrame frame);
```

Comment:
Underspecified: Javadoc doesn't say which "instrumented node" and doesn't say how long "before".  The node is the one mentioned in the provided `EventContext`.  How long "before" is a bit more subtle.  
Assume that the reader has the clear distinction between instrumentation and non-instrumentation nodes (see earlier comment requesting standard terminology for these two categories).  Then "before" in this context means "when the instrumented node is the next non-instrumentation node in the AST to be executed.  This make "before" precise with respect to the language execution, but remains silent with respect to ordering of any instrumentation nodes.  
This comment applies to the other two methods as well.